### ISSUE TITLE
semantics: SIGSEGV when handling unbounded array with a semantic

### ISSUE THREAD

**Issue by Keenuts** [microsoft/DirectXShaderCompiler#7582]
**Description**

https://godbolt.org/z/KGbfP9bGK

```hlsl
[shader("pixel")]
float main(float a[] : A) : SV_Target {
  return a[0];
}
```

Running in debug yields the following error instead.
```
Failed to allocate all input signature elements in available space.
UNREACHABLE executed at lib/HLSL/HLSignatureLower.cpp:523!
Aborted (core dumped)
```

**Actual Behavior**

Don't think this is legal HLSL, but DXC should not crash.

**Environment**
- DXC version: f94396ddffa8562a00d64a1db58d3f73f33b655a
- Host Operating System: Linux

**Comment by damyanp** [comment]
This is a real error that should be reported via a diagnostic rather than unreachable.

Repro that doesn't use unbounded arrays: https://godbolt.org/z/obf4KK84M
